Router and interface generation

adamthom-amzn opened the pull request:
*Description of changes:*

This lets me generate some interfaces and a router (mux?) that can be used in the app to determine what the user is trying to call.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

adamthom-amzn committed:
Add generation of server interfaces.

adamthom-amzn committed:
Add router generation for REST-JSON services.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        writer.writeInline("export type $LRequests = ", serviceInterfaceName);
+        for (Iterator<String> iter = requestInterfaces.iterator(); iter.hasNext();) {
+            writer.writeInline(iter.next());
+            if (iter.hasNext()) {
+                writer.writeInline(" | ");
+            }
+        }
```
I think `String.join(" | ", requestInterfaces)` would be much simpler here, unless there's some extra bit of magic I don't understand. Same for below.

adamthom-amzn replied:
Yeah, you're right, I was doing something different and then didn't adapt when that changed.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java:
```diff
@@ -214,6 +216,72 @@ public void generateResponseSerializers(GenerationContext context) {
+    public void generateRouter(GenerationContext context) {
+        TopDownIndex topDownIndex = TopDownIndex.of(context.getModel());
+
+        context.getWriter().addImport("restjson", null, "@aws-smithy/server-common");
```
This class isn't specific to json - it also is implemented by rest xml for example. I don't see anything below that's json specific, so this should just be renamed.

adamthom-amzn replied:
Makes sense.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java:
```diff
@@ -214,6 +216,72 @@ public void generateResponseSerializers(GenerationContext context) {
+    private void generateUriSpec(GenerationContext context,
+                                 OperationShape operation,
+                                 HttpTrait httpTrait) {
+        context.getWriter().openBlock("new restjson.UriSpec(", "),", () -> {
```
nit: you're calling `context.getWriter()` quite a bit. Calling it once and storing it would save some space

adamthom-amzn committed:
Add handler support for server generation.

This adds a generic handler concept, and then generates protocol-specific
handler factories for supported protocols.

adamthom-amzn commented:
I should have known that it would update this pr with my latest commit. Did not mean to put this much on you at once.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,142 @@
+        writer.addImport("HttpRequest", null, "@aws-sdk/protocol-http");
+        writer.addImport("HttpResponse", null, "@aws-sdk/protocol-http");
+
+        String serviceName = StringUtils.capitalize(service.getId().getName());
```
We should be using the symbol provider here, same for the names of the operations

adamthom-amzn replied:
In many of these cases I'm looking for the actual service or operation name for comparison against literal values. I don't want `<ServiceName>Client` or `<OperationName>Command` which is what the TS symbol provider names these things.

Even when I'm looking to just name something, I don't want to create `<ServiceName>ClientServiceHandler`. What's the path forward?

adamthom-amzn replied:
We talked about this offline: I'm going to look into creating a second symbol provider for SSDKs.

JordonPhillips replied:
To avoid bloating this too much, let's do that in a followup PR

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,142 @@
+        writer.openBlock("export class $LServiceHandler implements ServiceHandler {", "}", serviceName, () -> {
+            writer.write("private service: $LService;", serviceName);
+            writer.write("private mux: Mux<$S, $L>;", serviceName, operationsTypeName);
+            writer.write("private serializers: Record<$1L, OperationSerializer<$2LService[$1L]>>",
+                    operationsTypeName, serviceName);
```
there's no trailing semicolon

also, man TS is wild

adamthom-amzn replied:
yeah I thought this use of `Record` was particularly cool.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,142 @@
+                writer.write("requestHandler: new NodeHttpHandler(),");
+                writer.write("disableHostPrefix: true");
+            });
+            writer.openBlock("constructor(service: $1LService, "
+                            + "mux: Mux<$1S, $2L>, "
+                            + "serializers: Record<$2L, OperationSerializer<$1LService[$2L]>>) {", "}",
```
This is gonna need some doc comments in the generated TS. In particular it's not clear that `serializers` also includes deserializers. The `OperationSerializer` having both serializer and deserializer is a bit awkward. We've been using "serde" which also isn't the best, but at least doesn't give the wrong idea. `mux` is also a bit perplexing without some explanation.

adamthom-amzn replied:
`Serializer` is the only term I know for a thing that can both serialize and deserialize a type. A `Serdeer` is not a thing. I will always take naming suggestions, though.

Yes, me calling the interface Mux and the HTTPBinding implementation a Router isn't the best naming.

I'll try to clean this up.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,142 @@
+
+    private static void generateHandlerCase(TypeScriptWriter writer, OperationShape operation) {
+        String opName = operation.getId().getName();
+        writer.openBlock("case $S :", "", opName, () -> {
```
I wonder if we should handle the fully qualified name here too as a forwards compatibility deal. Probably not

adamthom-amzn replied:
Since this generated class is specific to one service, my hope was that fully qualifying anything was unnecessary.

JordonPhillips replied:
Yeah even if Smithy eventual had polymorphism it would be *really* weird for a service to have two operations with the same name.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,142 @@
+
+        String serviceInterfaceName = StringUtils.capitalize(service.getId().getName()) + "Service";
+
+        writer.openBlock("export interface $L {", "}", serviceInterfaceName, () -> {
+            for (OperationShape operation : operations) {
+                Symbol symbol = symbolProvider.toSymbol(operation);
+                writer.write("$L: $L<$T, $T>", StringUtils.capitalize(operation.getId().getName()),
+                        "__Operation",
+                        symbol.expectProperty("inputType", Symbol.class),
+                        symbol.expectProperty("outputType", Symbol.class));
+            }
+        });
```
My only problem with this is that it effectively forces you into having everything all in one place, so you couldn't use easily use apig to route to different lambdas for instance. Of course the counter to that is you can just do it yourself since it's not super hard, or just slap an `any` on so you don't even have to define the ones you don't want.

adamthom-amzn replied:
I think this interface is fine with all of the operations as required properties. I want to go back and make more of the generated TS reference this interface -- for instance, I use `keyof XService` in one place for a generic, but in another I generate a whole new union type from scratch.

The question is really does the ServiceHandler take an implementation of this interface, or does it take a `Partial` for the reason you describe? I think the answer is actually no. I think separating the deployment out into multiple lambdas is more of an infrastructure decision than a programming one. As it stands, you could write this one lambda with all of the operations implemented and have APIGateway route to the same lambda for each one or a different lambda for each one or a mix. It's all up to you.  You can still organize your code however you'd like (separate modules for different operations, one big file, whatever).

If I take an `Optional` then it feels more flexible, but in exchange it's easier to wire up a service that does not have a supplied implementation for a modeled operation.

JordonPhillips replied:
Yeah I think this is fine as is. In writing protocol tests against these I was able to `any` away any issues and I'm sure any customer could do that if they *really* want to, or otherwise wire things up themselves.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/ProtocolGenerator.java:
```diff
@@ -148,6 +148,30 @@ default void generateSharedComponents(GenerationContext context) {
+     */
+    void generateMux(GenerationContext context);
+
+    /**
+     * Generates the code used to determine the service and operation
+     * targeted by a given request.
+     *
+     * @param context Generation context.
+     */
+    void generateHandlerFactory(GenerationContext context);
+
+    /**
+     * Generates the code used to determine the service and operation
+     * targeted by a given request.
+     *
+     * @param context Generation context.
+     */
+    void generateHandlerFactory(GenerationContext context);
```
This is defined twice

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,152 @@
+            writer.openBlock("let input = await this.serializers.$L.deserialize(request, {", "});", opName, () -> {
+                writer.write("endpoint: () => Promise.resolve(request), ...this.serdeContextBase");
+            });
+            writer.write("let output = this.service.$L(input, request);", opName);
```
You can't use `let input` / `let output` in every case because for some reason each case doesn't have its own scope, and so if you have more than one operation you're technically re-declaring the variable. You can create a new scope by wrapping each case in `{}`.

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java:
```diff
@@ -214,6 +216,110 @@ public void generateResponseSerializers(GenerationContext context) {
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        String serviceName = StringUtils.capitalize(context.getService().getId().getName());
+
+        writer.addImport("BootlegService", null, "./server/interfaces");
+        writer.addImport("BootlegServiceHandler", null, "./server/handler");
```
These need to refer to the actual service/handler

adamthom-amzn replied:
whoops

JordonPhillips commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/ServerGenerator.java:
```diff
@@ -0,0 +1,152 @@
+    private static void generateHandlerCase(TypeScriptWriter writer, OperationShape operation) {
+        String opName = operation.getId().getName();
+        writer.openBlock("case $S :", "", opName, () -> {
+            writer.openBlock("let input = await this.serializers.$L.deserialize(request, {", "});", opName, () -> {
```
The output type of this is the union type and so it can't be passed into the service interface that expects a singular concrete type. You'll need to either `any` this or append `as ConcreteInput` or something.

adamthom-amzn replied:
I was wondering when having a model with just one test operation was going to bite me.

I think this actually points at the typing of the `Record` and `OperationSerializer` being not quite right. I am against `any`ing anything, I think I have a safe, but ugly temporary stopgap.

JordonPhillips replied:
Yeah I think the only other real option would be to abandon `Record`. Cool as it is, it's way more useful when you're not generating the code. A static mapping might ultimately be necessary, but that can be done in a followup.

JordonPhillips reviewed (approved):
There's a few follow up tasks, but this is good for now:

* introduce a server symbol provider
* look into reworking record into something more explicit
* adding configuration

JordonPhillips committed:
Merge pull request #270 from JordonPhillips/server-serde

Add basic support for server serde in http binding protocols

adamthom-amzn commented:
Closing this in favor of https://github.com/awslabs/smithy-typescript/pull/278 which is basically the same
